src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs

```diff
@@ -184,13 +184,17 @@ public void Serialize(ObjectWriter writer, CancellationToken cancellationToken)
             }
 
             var valuesBuilder = new SortedDictionary<OptionKey, (OptionValueKind, object?)>(OptionKeyComparer.Instance);
-            foreach (var (optionKey, value) in _serializableOptionValues)
+            foreach (var optionKey in _serializableOptionValues.Keys.Union(_changedOptionKeys))
             {
-                Debug.Assert(ShouldSerialize(optionKey));
+                object? value;
 
-                if (!_serializableOptions.Contains(optionKey.Option))
```

Comment:
I am wondering if we can completely get rid of `_serializableOptions` now and change `_serializableOptionValues` to now hold serializable + changed values, latter being always considered serializable. The else part below can be completely deleted: https://github.com/dotnet/roslyn/blob/8ab475f5aa31e6dc9093ff720a1cd251576b0fbd/src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs#L130-L139

Reply:
What was the original intent behind the split? Was the idea that non-serializable stuff we should just keep in the workspace option set so we avoid accidentally trying to serialize it and fail?

Reply:
Yes, that is correct - I thought we could have clear separation between serializable part that is in solution snapshot and the host/workspace specific part.